### PR TITLE
Add Michelson Golden Tests

### PR DESCRIPTION
**Pull request by Acentelles**
- Add Michelson golden tests
- Allow golden tests to be compared with their Show instance
- Add tests for `main` having different signature

### CODE REVIEW

**Comment by Acentelles** [comment]
Can we merge this @mariari ?

**Review by mariari — APPROVED** [review]
Eeverything looks good

File: library/Backends/Michelson/test/Test/Golden.hs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+-- | Discover golden tests for input files with extension @.ju@ and output
+-- files with extension @.michelson@.
+discoverGoldenTestsCompile ::
+  -- | the directory in which to recursively look for golden tests
+  FilePath ->
+  IO TestTree
+discoverGoldenTestsCompile (withJuvixRootPath -> p) = discoverGoldenTests [".ju"] ".michelson" getGolden (expectSuccess . compile) p
+  where
+    compile file = Michelson.compileMichelson  =<< typecheck file
```
Comment (rokopt) [inline library/Backends/Michelson/test/Test/Golden.hs]: Same newline comment as above.

File: library/Backends/Michelson/test/Test/Golden.hs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        discoverGoldenTestsTypecheck "test/examples/negative/michelson"
+      ]
+
+-- | Discover golden tests for input files with extension @.ju@ and output
+-- files with extension @.typecheck@.
+discoverGoldenTestsTypecheck ::
+  -- | the directory in which to recursively look for golden tests
+  FilePath ->
+  IO TestTree
+discoverGoldenTestsTypecheck (withJuvixRootPath -> p) = discoverGoldenTests [".ju"] ".typecheck" getGolden (expectSuccess . typecheck) p
```
Comment (rokopt) [inline library/Backends/Michelson/test/Test/Golden.hs]: Don't we need `expectFailure` for the tests in `negative`?  Or am I misunderstanding how the golden tests work?

File: library/Backends/Michelson/test/Test/Golden.hs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  context <- Pipeline.parseWithLibs (withJuvixRootPath <$> libs) Michelson.BMichelson contract
+  Pipeline.typecheck @Michelson.BMichelson context
+
+-- | Discover golden tests for input files with extension @.ju@ and output
+-- files with extension @.michelson@.
+discoverGoldenTestsCompile ::
+  -- | the directory in which to recursively look for golden tests
+  FilePath ->
+  IO TestTree
+discoverGoldenTestsCompile (withJuvixRootPath -> p) = discoverGoldenTests [".ju"] ".michelson" getGolden (expectSuccess . compile) p
```
Comment (rokopt) [inline library/Backends/Michelson/test/Test/Golden.hs]: Same `expectFailure` comment as above.

File: library/Backends/Michelson/test/Test/Golden.hs (hunk 2 above)
Comment (rokopt) [reply library/Backends/Michelson/test/Test/Golden.hs]: Oh -- I see this was fixed in a later commit.  Would you please squash and re-divide commits to eliminate reviews of obsolete code, and also rebase to remove merge commits from the history, prior to review (for future PRs)?

File: library/Pipeline/test/Test/RecGroups.hs [hunk 4]
```diff
@@ -37,8 +37,8 @@ pipeline =
           Right c <-
             Pipeline.toCore
               ( withJuvixExamplesPath
-                  <$> [ "positive/michelson/rec-groups/Rec-Groups.ju",
-                        "positive/michelson/rec-groups/Rec-Groups-Helper.ju"
+                  <$> [ "to-fix/rec-groups/Rec-Groups.ju",
+                        "to-fix/rec-groups/Rec-Groups-Helper.ju"
```
Comment (rokopt) [inline library/Pipeline/test/Test/RecGroups.hs:41]: Are these tests broken by some incompleteness in Juvix at the moment?  It'd be nice to have a comment for any `to-fix` tests explaining what needs to be implemented before they can be moved to `positive` or `negative`.

File: library/Backends/Michelson/src/Juvix/Backends/Michelson/Pipeline.hs [hunk 5]
```diff
@@ -27,13 +27,14 @@ instance HasBackend BMichelson where
...
 compileMichelson term = do
-    let (res, _logs) = M.compileContract $ CorePipeline.toRaw term
-    case res of
-      Right c -> pure $ M.untypedContractToSource (fst c)
-      Left err -> Feedback.fail $ show err
\ No newline at end of file
+  let (res, _logs) = M.compileContract $ CorePipeline.toRaw term
+  case res of
+    Right c -> pure $ M.untypedContractToSource (fst c)
+    Left err -> Feedback.fail $ show err
```
Comment (rokopt) [inline library/Backends/Michelson/src/Juvix/Backends/Michelson/Pipeline.hs]: Commit-hook changes like this would be good to squash too.

**Review by rokopt — APPROVED** [review]

File: library/Pipeline/test/Test/RecGroups.hs (hunk 4 above)
Comment (Acentelles) [reply library/Pipeline/test/Test/RecGroups.hs:41]: That's a good point. At the moment, they are backend specific. So they fail for every backend, but they work for this bit of the pipeline. I'll think about a better folder structure to encapsulate all these nuances. Any suggestions welcome!

File: library/Backends/Michelson/test/Test/Golden.hs (hunk 2 above)
Comment (Acentelles) [reply library/Backends/Michelson/test/Test/Golden.hs]: Yes, will do. Thanks for spotting it

Comment (Acentelles) [reply library/Backends/Michelson/test/Test/Golden.hs]: Yeah, weird you reviewed obsolete code. I've just seen it's fixed in a later commit
